main/src/vocab/nllb_vocab.rs

```diff
@@ -0,0 +1,245 @@
+            ..Default::default()
+        };
+
+        Self::from_values_and_special_token_map(tokenizer.model.vocab, special_config)
+    }
+
+    fn token_to_id(&self, token: &str) -> i64 {
+        self._token_to_id(
+            token,
+            self.values(),
+            self.special_values(),
+            self.get_unknown_value(),
+        )
+    }
+
+    fn id_to_token(&self, id: &i64) -> String {
+        self._id_to_token(
+            id,
+            self.indices(),
+            self.special_indices(),
+            self.get_unknown_value(),
+        )
+    }
+
+    fn from_file_with_special_token_mapping<P: AsRef<Path>, S: AsRef<Path>>(
+        path: P,
+        special_token_mapping_path: S,
+    ) -> Result<Self, TokenizerError>
+    where
+        Self: Sized,
+    {
+        let tokenizer_config = Tokenizer::deserialize(path)?;
+        let special_config = read_special_token_mapping_file(special_token_mapping_path)?;
+        Self::from_values_and_special_token_map(tokenizer_config.model.vocab, special_config)
+    }
+
+    fn from_values_and_special_token_map(
+        values: HashMap<String, i64>,
+        special_token_map: SpecialTokenMap,
+    ) -> Result<Self, TokenizerError>
+    where
+        Self: Sized,
+    {
+        let mut result = Self {
+            values,
+            indices: HashMap::new(),
+            special_values: HashMap::new(),
+            special_indices: HashMap::new(),
+            language_codes_bytes: HashSet::new(),
+            special_token_storage: special_token_map,
+        };
+
+        let mut special_values = HashMap::new();
+        let mut language_code_bytes = HashSet::new();
+
+        let mut reserve_special =
+            |t| register_as_special_value(t, &result.values, &mut special_values);
+
+        reserve_special(result.get_bos_value())?;
+        reserve_special(result.get_eos_value())?;
+        reserve_special(result.get_sep_value())?;
+        reserve_special(result.get_pad_value())?;
+        reserve_special(result.get_unknown_value())?;
+
+        if let Some(languages) = result
+            .special_token_storage
+            .additional_special_tokens
+            .as_ref()
+        {
+            for language in languages {
+                reserve_special(language)?;
+                language_code_bytes.insert(language.as_bytes().to_vec());
+            }
+        } else {
+            for language in EXTENDED_FAIRSEQ_LANGUAGE_CODES {
+                reserve_special(language)?;
+                language_code_bytes.insert(language.as_bytes().to_vec());
+            }
+        }
+
+        let indices = swap_key_values(&result.values);
+        let special_indices = swap_key_values(&special_values);
+        result.indices = indices;
+        result.special_indices = special_indices;
+        result.special_values = special_values;
+        result.language_codes_bytes = language_code_bytes;
+        Ok(result)
+    }
+}
+
+#[derive(Deserialize)]
+struct Model {
```

Comment:
Looks like this should be removed

Reply:
This code is required for `tokenizer.json` deserialization (it nested and contains vocabulary + extra meta information).

Reply:
Please see the comment above. I think creating a `NLLBSpecialTokenMap` and `From<NLLBSpecialTokenMap>` for `SpecialTokenMap` would help keeping the original `SpecialTokenMap` unchanged and remove the need for a separate deserialization `Model` struct. What do you think?

Reply:
Just double-checked the file and you are correct - apologies did not realize that NLLB had a different vocabulary format altogether.

Reply:
Workaround was added in https://github.com/npatsakula/rust-tokenizers/pull/5.